examples/ERC721/index.rsh

```diff
@@ -1,2 +1,307 @@
+          //check(isSome(m_bal), "ERC721::balanceOf: No balance for address");
+          return fromSome(m_bal, 0);
+        });
+        const tokenExists = (tokenId) => isSome(owners[tokenId]);
+
+        const ownerOf = (tokenId) => {
+          check(tokenExists(tokenId), "ERC721::ownerOf: Owner query for non-existent token");
+          const m_owner = owners[tokenId];
+          return fromSome(m_owner, zeroAddr);
+        };
+        V.ownerOf.set(ownerOf);
+
+        const getApproved = (tokenId) => {
+          check(tokenExists(tokenId), "ERC721::getApproved: approved query for non-existent token")
+          const m_approval = tokenApprovals[tokenId];
+          return fromSome(m_approval, zeroAddr);
+        };
+        V.getApproved.set(getApproved);
+
+        const isApprovedForAll = (owner, operator) => {
+          const m_approved = operatorApprovals[[owner, operator]];
+          return fromSome(m_approved, false);
+        };
+        V.isApprovedForAll.set(isApprovedForAll);
+
+        const isApprovedOrOwner = (spender, tokenId) => {
+          check(tokenExists(tokenId), "isApprovedOrOwner: token exists");
+          const owner = ownerOf(tokenId);
+          return spender == owner || isApprovedForAll(owner, spender) || getApproved(tokenId) == spender;
+        }
+
+        V.tokenURI.set((tokenId) => {
+          check(tokenExists(tokenId), "tokenURI: URI query for non-existent token");
+          return StringDyn.concat(tokenURI, StringDyn(tokenId));
+        });
+
+      })
+      .while(true)
+      .invariant(balance() == 0)
+      .define(() => {
+        const approve = (to, tokenId) => {
+          const owner = ownerOf(tokenId);
+          tokenApprovals[tokenId] = to;
+          E.Approval(owner, to, tokenId);
+        }
+        const transferChecks = (from_, to, tokenId) => {
+          const owner = ownerOf(tokenId);
+          check(owner == from_, "ERC721::transfer: transfer from incorrect owner");
+          check(to != zeroAddr, "ERC721::transfer: transfer to the zero address");
+        }
+        const transfer_ = (from_, to, tokenId) => {
+          transferChecks(from_, to, tokenId);
+          approve(zeroAddr, tokenId);
+          balances[from_] = maybe(balances[from_], 0, (b) => b - 1);
+          balances[to]    = maybe(balances[to]   , 1, (b) => b + 1);
+          owners[tokenId] = to;
+          E.Transfer(from_, to, tokenId);
+        }
+      })
+      .api_(I.safeTransferFrom1, (from_, to, tokenId, data) => {
+        check(isApprovedOrOwner(this, tokenId), "ERC721::safeTransferFrom: transfer caller is not owner nor approved");
+        transferChecks(from_, to, tokenId);
+        return [ (k) => {
+          transfer_(from_, to, tokenId);
+          // TODO - this should be a remote call if and only if the to address is a contract.  We need to use the future feature Contract.fromAddress here to determine that.
```

Comment:
I believe that this function only makes sense when `to` is a Contract, so we should just change the type, and then maybe it is not `Contract.fromAddress` that we want but `Address.fromContract` that we want